In [57]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import country_converter as coco
import warnings
warnings.filterwarnings('ignore')

In [58]:
original_df = pd.read_csv("data/stackoverflow_users_all_notext.csv", nrows=10000)

In [59]:
"""Show the data."""
display(original_df[130:140])

id  age                                        location  \
130  4845736  NaN                                             NaN   
131  4890609  NaN  Belo Horizonte - State of Minas Gerais, Brazil   
132  4965490  NaN                                    Talca, Chile   
133  5006338  NaN                                             NaN   
134  5073970  NaN                                             NaN   
135  5150946  NaN                            Melbourne, Australia   
136  5185772  NaN                                             NaN   
137  5231705  NaN                                         jakarta   
138  5248405  NaN                                             NaN   
139  5281835  NaN                                             NaN   

                   creation_date  reputation  up_votes  down_votes  
130   2015-04-29 09:41:03.96 UTC           1         0           0  
131   2015-05-12 09:03:06.03 UTC           1         0           0  
132   2015-06-02 13:36:06.94 UTC           1         0           0  
133  2015-06-13 14:17:20.823 UTC           1         0           0  
134  2015-07-02 13:15:12.983 UTC           1         0           0  
135   2015-07-24 05:58:04.17 UTC           1         0           0  
136  2015-08-03 13:10:44.243 UTC           1         0           0  
137  2015-08-16 06:01:09.437 UTC           1         0           0  
138  2015-08-20 16:12:00.633 UTC           1         0           0  
139  2015-08-30 12:45:03.383 UTC           1         0           0

The location field seems to have the country name as the last part quite consistently, separated 
by a comma and space. Some of the countries are, however, named using some non-English form. 
Let's download a dataset of country names in different languages and use them to match locations to countries.

In [60]:
clean_df = original_df[original_df.location.notnull()]
notinteresting_df = original_df[original_df.location.isnull()]

In [61]:
"""Extract the country field out of the location field"""
clean_df["country"] = clean_df["location"].apply(lambda x: x.split(", ")[-1].lower() if isinstance(x,str) else np.nan)

In [62]:
clean_df.head(5)

id  age                           location  \
1  7019478  NaN                          Singapore   
2  7074992  NaN                              Egypt   
3  7093392  NaN        New York, NY, United States   
6  7395536  NaN  Cluj-Napoca, Cluj County, Romania   
8  7497308  NaN                        Bodø, Norge   

                 creation_date  reputation  up_votes  down_votes  \
1   2016-10-14 13:36:30.96 UTC           1         0           0   
2   2016-10-26 12:11:08.41 UTC           1         0           0   
3   2016-10-31 02:19:04.64 UTC           1         0           0   
6  2017-01-09 16:36:43.747 UTC           1         0           0   
8  2017-01-31 21:13:26.053 UTC           1         0           0   

         country  
1      singapore  
2          egypt  
3  united states  
6        romania  
8          norge

In [63]:
"""Load the dataset of country names in different languages got from ip2location.com."""
country_names = pd.read_csv("data/IP2LOCATION-COUNTRY-MULTILINGUAL/IP2LOCATION-COUNTRY-MULTILINGUAL.CSV")

In [64]:
display(country_names.head())

LANG  LANG_NAME COUNTRY_ALPHA2_CODE COUNTRY_ALPHA3_CODE  \
0   AF  AFRIKAANS                  AD                 AND   
1   AF  AFRIKAANS                  AE                 ARE   
2   AF  AFRIKAANS                  AF                 AFG   
3   AF  AFRIKAANS                  AG                 ATG   
4   AF  AFRIKAANS                  AI                 AIA   

   COUNTRY_NUMERIC_CODE                COUNTRY_NAME  
0                    20                     Andorra  
1                   784  Verenigde Arabiese Emirate  
2                     4                  Afganistan  
3                    28          Antigua en Barbuda  
4                   660                    Anguilla

In [65]:
"""Pick only the relevant part of the data"""
country_name_to_code = country_names[['COUNTRY_NAME','COUNTRY_ALPHA3_CODE']]
country_name_to_code.COUNTRY_NAME = country_name_to_code.COUNTRY_NAME.str.lower()
"""Drop duplicates since many (or probably all) countries have the same name in several languages"""
country_name_to_code = country_name_to_code[~country_name_to_code.COUNTRY_NAME.duplicated()]

In [66]:
clean_df = clean_df.merge(country_name_to_code, left_on="country", right_on="COUNTRY_NAME", how="left")
clean_df.drop(["COUNTRY_NAME"], axis=1, inplace=True)

In [67]:
clean_df.head()

id  age                           location  \
0  7019478  NaN                          Singapore   
1  7074992  NaN                              Egypt   
2  7093392  NaN        New York, NY, United States   
3  7395536  NaN  Cluj-Napoca, Cluj County, Romania   
4  7497308  NaN                        Bodø, Norge   

                 creation_date  reputation  up_votes  down_votes  \
0   2016-10-14 13:36:30.96 UTC           1         0           0   
1   2016-10-26 12:11:08.41 UTC           1         0           0   
2   2016-10-31 02:19:04.64 UTC           1         0           0   
3  2017-01-09 16:36:43.747 UTC           1         0           0   
4  2017-01-31 21:13:26.053 UTC           1         0           0   

         country COUNTRY_ALPHA3_CODE  
0      singapore                 SGP  
1          egypt                 EGY  
2  united states                 USA  
3        romania                 ROU  
4          norge                 NOR

In [68]:
print("Proportion of matches: ", sum(clean_df.COUNTRY_ALPHA3_CODE.notnull())/sum(clean_df.location.notnull()))

Proportion of matches:  0.8127413127413128


The method manages to add a country tag to 86.9% of the users who have informed a country in the first place. Let's then look at the users that weren't matched.

In [69]:
clean_df[clean_df.location.notnull() & clean_df.COUNTRY_ALPHA3_CODE.isnull()].head()

id  age             location                creation_date  \
6   7579333  NaN       Beijing, 北京市中国  2017-02-17 07:27:20.823 UTC   
13  8448975  NaN            Bangalore   2017-08-11 05:32:54.82 UTC   
15  8623520  NaN           deverloper  2017-09-17 19:52:14.057 UTC   
20  9569837  NaN                   UK  2018-03-29 12:56:51.147 UTC   
22  9867362  NaN  Hangzhou, 浙江省 China   2018-05-30 01:06:04.59 UTC   

    reputation  up_votes  down_votes     country COUNTRY_ALPHA3_CODE  
6            1         0           0       北京市中国                 NaN  
13           1         0           0   bangalore                 NaN  
15           1         0           0  deverloper                 NaN  
20           1         0           0          uk                 NaN  
22           1         0           0   浙江省 china                 NaN

Seems like a part of the problem is people giving out only the city or some funny stuff. Let's try addressing this by loading a dataset of city names and trying to match them with the remaining, unmatched data.

In [70]:
world_cities = pd.read_csv("data/worldcities.csv")
display(world_cities.head())

city   city_ascii      lat      lng country iso2 iso3   admin_name  \
0    Malishevë    Malisheve  42.4822  20.7458  Kosovo   XK  XKS    Malishevë   
1      Prizren      Prizren  42.2139  20.7397  Kosovo   XK  XKS      Prizren   
2  Zubin Potok  Zubin Potok  42.9144  20.6897  Kosovo   XK  XKS  Zubin Potok   
3     Kamenicë     Kamenice  42.5781  21.5803  Kosovo   XK  XKS     Kamenicë   
4         Viti         Viti  42.3214  21.3583  Kosovo   XK  XKS         Viti   

  capital  population          id  
0   admin         NaN  1901597212  
1   admin         NaN  1901360309  
2   admin         NaN  1901608808  
3   admin         NaN  1901851592  
4   admin         NaN  1901328795

In [71]:
display(world_cities[world_cities.city=="Moscow"])

city city_ascii      lat       lng        country iso2 iso3  \
2220   Moscow     Moscow  55.7522   37.6155         Russia   RU  RUS   
11641  Moscow     Moscow  46.7307 -116.9986  United States   US  USA   

      admin_name  capital  population          id  
2220      Moskva  primary  10452000.0  1643318494  
11641      Idaho      NaN     25579.0  1840019868

There are cities with the same names in many countries. Let's assume that if no country was said, it's probably the largest city.
One could in principle also use the administrative area name here but it doesn't seem as likely that people would put just the name of the administrative area but not the country or city name so let's try the city first.

In [72]:
world_cities = world_cities.sort_values(by="population").drop_duplicates(subset="city",keep="last").sort_index()

In [73]:
"""Again, take only interesting parts"""
city_to_country_code = world_cities[['city','iso3']]
city_to_country_code.city = city_to_country_code.city.str.lower()

In [74]:
clean_df["city_candidate"] = clean_df["location"].apply(lambda x: x.split(", ")[0].lower() if isinstance(x,str) else np.nan)

In [75]:
clean_df = clean_df.merge(city_to_country_code, left_on="city_candidate", right_on="city", how="left")
clean_df["COUNTRY_ALPHA3_CODE"][clean_df["COUNTRY_ALPHA3_CODE"].isnull()] = clean_df["iso3"][clean_df["COUNTRY_ALPHA3_CODE"].isnull()]

In [76]:
print("Proportion of matched: ", sum(clean_df.COUNTRY_ALPHA3_CODE.notnull())/sum(clean_df.location.notnull()))

Proportion of matched:  0.9146718146718147


Seems like using the city name brings us up to 92.3% of matches.

In [77]:
"""Clean out some useless fields"""
clean_df.drop(["city_candidate", "city", "iso3","country"],axis=1, inplace=True)

In [78]:
"""See the remaining users that are still unmatched. Also prepare a lowercase location column for further analysis."""
clean_df["location_lc"] = clean_df.location.str.lower()
remaining = clean_df[clean_df.location.notnull()][clean_df.COUNTRY_ALPHA3_CODE.isnull()]
display(remaining[100:110])

id  age                                           location  \
1267  3562541  NaN                                          Minnesota   
1272  5749992  NaN                         Somewhere Over the Rainbow   
1280  8821565  NaN  Taipei 101, Section 5, Xinyi Road, Xinyi Distr...   
1282  4366341  NaN                                               Reek   
1286  4593809  NaN                                           Hull, UK   
1287  7570131  NaN  Walchand College of Engineering ,sangli-416415...   
1295  1792154  NaN                                              Idaho   
1296  2299708  NaN                                                 UT   
1303  2388579  NaN                                        Studio City   
1311  3330609  NaN                                             Online   

                    creation_date  reputation  up_votes  down_votes  \
1267  2014-04-23 00:39:46.953 UTC           3         0           0   
1272  2016-01-05 23:47:53.463 UTC           3         0           0   
1280  2017-10-23 20:07:19.247 UTC           3         0           0   
1282   2014-12-16 12:33:32.24 UTC           3         0           0   
1286  2015-02-22 13:56:59.387 UTC           3         0           0   
1287  2017-02-15 16:22:29.603 UTC           3         0           0   
1295   2012-11-01 17:49:20.17 UTC           3         0           0   
1296   2013-04-19 14:55:46.24 UTC           3         0           0   
1303  2013-05-16 06:00:49.067 UTC           3         0           0   
1311  2014-02-20 00:48:26.767 UTC           3         0           0   

     COUNTRY_ALPHA3_CODE                                        location_lc  
1267                 NaN                                          minnesota  
1272                 NaN                         somewhere over the rainbow  
1280                 NaN  taipei 101, section 5, xinyi road, xinyi distr...  
1282                 NaN                                               reek  
1286                 NaN                                           hull, uk  
1287                 NaN  walchand college of engineering ,sangli-416415...  
1295                 NaN                                              idaho  
1296                 NaN                                                 ut  
1303                 NaN                                        studio city  
1311                 NaN                                             online

Seems like the remaining problem is in part that the format isn't necessarily
always city, admin area, country but sometimes some other ordering. Also, for instance UK is a common abbreviation for the United Kingdom but it wasn't in the country dataset. A regex search for the rest seems like a good idea.

In [79]:
country_name_to_code_2 = country_name_to_code.append([{"COUNTRY_NAME" : "uk", "COUNTRY_ALPHA3_CODE" : "GBR"},
                                                     {"COUNTRY_NAME" : "usa", "COUNTRY_ALPHA3_CODE" : "USA"},
                                                     {"COUNTRY_NAME" : "south korea", "COUNTRY_ALPHA3_CODE" : "KOR"}] ,ignore_index=True)

In [ ]:
i = 0
for idx, row in country_name_to_code_2.iterrows():
    country_name,iso3 = row[0],row[1]
    matching = remaining.location_lc.str.contains("\\b"+country_name+"\\b")
    remaining.COUNTRY_ALPHA3_CODE[matching] = iso3
    if i % 1000 == 0:
        print(idx/len(country_name_to_code)*100, "% calculated")
    i += 1

0.0 % calculated
7.471049682480388 % calculated
14.942099364960775 % calculated


In [26]:
clean_df = clean_df.merge(remaining.drop(["id","location","creation_date","reputation","up_votes","down_votes","location_lc","age"],axis=1), left_index=True, right_index=True, how="left")
clean_df.COUNTRY_ALPHA3_CODE_x[clean_df.COUNTRY_ALPHA3_CODE_x.isnull()] = clean_df.COUNTRY_ALPHA3_CODE_y[clean_df.COUNTRY_ALPHA3_CODE_x.isnull()]

In [27]:
clean_df.drop(["COUNTRY_ALPHA3_CODE_y"], axis=1, inplace=True)
clean_df.rename(columns={"COUNTRY_ALPHA3_CODE_x" : "country_iso3"}, inplace=True)

In [28]:
print("Proportion of matched: ", sum(clean_df.country_iso3.notnull())/sum(clean_df.location.notnull()))

Proportion of matched:  0.9474566236164214


Using the regex method brings the results already to 94.75 percent. Let's look at the remaining issues.

In [29]:
clean_df[clean_df.location.notnull() & clean_df.country_iso3.isnull()][160:180]

id  age                            location  \
8380   9333556  NaN                             Czechia   
8409   9719937  NaN                           Worldwide   
8538   5991722  NaN   Hranice na Moravě, Hranice, Česko   
8622   6292000  NaN                           Somewhere   
8732   6783798  NaN                         Schweinfurt   
8800   7165275  NaN                             Vermont   
8901   7771928  NaN                            Bangaore   
8924   7878102  NaN          Nagrota Bypass Road, Jammu   
8931   7898657  NaN               中国河北省Shijiazhuang Shi   
8947   7974340  NaN                                U.S.   
9004   8377142  NaN                       Beroun, Česko   
9007   8398531  NaN                               Korea   
9031   8580067  NaN               Lviv, Lviv Oblast, UA   
9074   8998175  NaN                               @home   
9180   9819333  NaN                               Noida   
9215  10306949  NaN                                  US   
9291   5971239  NaN  San Francisco Bay Area, California   
9294   6004645  NaN                     Wisconsin, U.S.   
9425   6589588  NaN         Charlotte · Bremen · Taipei   
9491   6820646  NaN          Somewhere over the rainbow   

                    creation_date  reputation  up_votes  down_votes  \
8380  2018-02-08 14:04:47.567 UTC          30         0           0   
8409   2018-04-30 06:38:27.79 UTC          30        47           0   
8538   2016-02-27 22:37:37.31 UTC         461        27          17   
8622  2016-05-04 17:01:26.607 UTC         308       693           0   
8732   2016-09-01 14:29:22.53 UTC        1061       129          26   
8800  2016-11-16 03:00:26.087 UTC        1943        60          12   
8901  2017-03-27 04:26:53.803 UTC          37         0           0   
8924   2017-04-17 10:18:59.57 UTC          61         8           0   
8931  2017-04-21 00:05:31.497 UTC         431         4           3   
8947  2017-05-06 20:41:23.647 UTC         126        12           0   
9004  2017-07-27 15:14:26.397 UTC          52         3           0   
9007  2017-08-01 08:41:32.603 UTC          15         1           0   
9031   2017-09-08 12:36:22.56 UTC          18         0           0   
9074  2017-11-23 20:06:40.347 UTC         149        15           2   
9180   2018-05-20 15:10:25.88 UTC         802        10           0   
9215  2018-09-02 15:19:39.513 UTC         111         5           0   
9291   2016-02-23 22:15:53.26 UTC        1184        45           1   
9294   2016-03-01 20:53:43.71 UTC          24         6           0   
9425   2016-07-14 13:04:20.55 UTC         328         1           0   
9491  2016-09-12 01:43:37.047 UTC         154        61           0   

     country_iso3                         location_lc  
8380          NaN                             czechia  
8409          NaN                           worldwide  
8538          NaN   hranice na moravě, hranice, česko  
8622          NaN                           somewhere  
8732          NaN                         schweinfurt  
8800          NaN                             vermont  
8901          NaN                            bangaore  
8924          NaN          nagrota bypass road, jammu  
8931          NaN               中国河北省shijiazhuang shi  
8947          NaN                                u.s.  
9004          NaN                       beroun, česko  
9007          NaN                               korea  
9031          NaN               lviv, lviv oblast, ua  
9074          NaN                               @home  
9180          NaN                               noida  
9215          NaN                                  us  
9291          NaN  san francisco bay area, california  
9294          NaN                     wisconsin, u.s.  
9425          NaN         charlotte · bremen · taipei  
9491          NaN          somewhere over the rainbow

Seems like the remaining issues are mostly people giving out only the city or administrative area name in some format that wasn't considered before. Let's try to fix this with a similar regex matching for both.

In [30]:
world_cities_2 = pd.read_csv("data/worldcities.csv")
world_cities_2 = world_cities.sort_values(by="population").drop_duplicates(subset="admin_name",keep="last").sort_index()
admin_to_country_code = world_cities_2[['admin_name','iso3']]
admin_to_country_code.admin_name = admin_to_country_code.admin_name.str.lower()

In [32]:
"""Weird thing: NaN value in one of the admin names"""
remaining = clean_df[clean_df.location_lc.notnull()][clean_df.country_iso3.isnull()]
i = 0
for idx, row in admin_to_country_code.iterrows():
    admin_name,iso3 = row[0],row[1]
    matching = remaining.location_lc.str.contains("\\b"+str(admin_name)+"\\b")
    remaining.country_iso3[matching] = iso3
    if i % 1000 == 0:
        print(i/len(admin_to_country_code)*100, "% calculated")
    i += 1

0.0 % calculated
25.73340195573855 % calculated
51.4668039114771 % calculated
77.20020586721564 % calculated


In [33]:
remaining.rename(columns={"country_iso3" : "rem_iso3"}, inplace=True)

In [34]:
clean_df = clean_df.merge(remaining.drop(["id","location","creation_date","reputation","up_votes","down_votes","location_lc","age"],axis=1), left_index=True, right_index=True, how="left")
clean_df.country_iso3[clean_df.country_iso3.isnull()] = clean_df.rem_iso3[clean_df.country_iso3.isnull()]
clean_df.drop(["rem_iso3"], axis=1, inplace=True)

In [35]:
remaining = clean_df[clean_df.location_lc.notnull() & clean_df.country_iso3.isnull()]
i = 0
for idx, row in city_to_country_code.iterrows():
    city_name,iso3 = row[0],row[1]
    matching = remaining.location_lc.str.contains("\\b"+city_name+"\\b")
    remaining.country_iso3[matching] = iso3
    if i % 1000 == 0:
        print(i/len(city_to_country_code)*100, "% calculated")
    i += 1

0.0 % calculated
8.634832915983075 % calculated
17.26966583196615 % calculated
25.90449874794923 % calculated
34.5393316639323 % calculated
43.17416457991538 % calculated
51.80899749589846 % calculated
60.44383041188153 % calculated
69.0786633278646 % calculated
77.71349624384769 % calculated
86.34832915983075 % calculated
94.98316207581384 % calculated


In [36]:
remaining.rename(columns={"country_iso3" : "rem_iso3"}, inplace=True)

In [37]:
clean_df = clean_df.merge(remaining.drop(["id","location","creation_date","reputation","up_votes","down_votes","location_lc","age"],axis=1), left_index=True, right_index=True, how="left")
clean_df.country_iso3[clean_df.country_iso3.isnull()] = clean_df.rem_iso3[clean_df.country_iso3.isnull()]
clean_df.drop(["rem_iso3"], axis=1, inplace=True)

In [38]:
display(admin_to_country_code[[not isinstance(x,str) for x in admin_to_country_code.admin_name]])

admin_name iso3
7818        NaN  JEY

Funny fact: before we needed to have str(admin_name) in the code because there's a place called Nan in Thailand. 

In [39]:
sum(clean_df.country_iso3.notnull())/sum(clean_df.location.notnull())

0.956600201369446

With the final method we get 95.66% of country matches to locations.

In [40]:
clean_df.drop(["location","location_lc"],axis=1,inplace=True)

In [41]:
clean_df.to_csv("data/stackoverflow_users_2016-2019_notext_clean.csv")

In [42]:
original_df[clean_df.country_iso3.isnull() & original_df.location.notnull()]

id  age                              location  \
32        5876676  NaN                    中国广东省Guangzhou Shi   
49        5924453  NaN                              Hobbiton   
244       7008814  NaN                                  mars   
425       7878292  NaN                          Arandjelovac   
435       7945249  NaN                             127.0.0.1   
...           ...  ...                                   ...   
6035679   6142801  NaN             New Jersey, United States   
6035680   9703547  NaN  Yokohama, Kanagawa Prefecture, Japan   
6035682  11258913  NaN                      Mississippi, USA   
6035685   9862136  NaN                       Everett, 워싱턴 미국   
6035687  11636649  NaN          Nagercoil, Tamil Nadu, India   

                       creation_date  reputation  up_votes  down_votes  \
32       2016-02-03 07:47:09.317 UTC          92        86           0   
49       2016-02-14 06:50:46.743 UTC         610        40           5   
244      2016-10-12 15:28:03.683 UTC          15        68           0   
425        2017-04-17 11:10:19.7 UTC          18         3           0   
435       2017-04-30 22:50:32.76 UTC         154         7           0   
...                              ...         ...       ...         ...   
6035679  2016-04-01 01:57:26.287 UTC          31         0           0   
6035680  2018-04-26 10:47:03.497 UTC          31         2           0   
6035682  2019-03-26 07:21:08.673 UTC          31        24           0   
6035685  2018-05-29 04:55:22.243 UTC          31        10           0   
6035687  2019-06-12 12:03:40.373 UTC          31        28           0   

                    country  
32       中国广东省guangzhou shi  
49                 hobbiton  
244                    mars  
425            arandjelovac  
435               127.0.0.1  
...                     ...  
6035679       united states  
6035680               japan  
6035682                 usa  
6035685              워싱턴 미국  
6035687               india  

[1228364 rows x 8 columns]

In [43]:
original_df[original_df.location.notnull() & clean_df.country_iso3.isnull()][100:130]

id  age                                           location  \
6163   8644091  NaN                                         Babu Nagar   
6174   8813176  NaN                                     Mariana Trench   
6180   8867531  NaN                                            Czechia   
6203   9015220  NaN                                             Korean   
6211   9061350  NaN                                        Malu abrupt   
6262   9537794  NaN                                         Everywhere   
6309   9975719  NaN                                                 DC   
6327  10211923  NaN                                          HumanLand   
6356   5789789  NaN                                        GetPosition   
6446   6176383  NaN                                          Milky-Way   
6508   6440619  NaN                                          Somewhere   
6575   6706654  NaN                                           Internet   
6603   6821443  NaN                                              中国上海市   
6689   7358586  NaN                                           The moon   
6788   7924725  NaN                                             Europe   
6819   8182375  NaN                                             Ismail   
6860   8496457  NaN                                             Europa   
6951   9188006  NaN                                          The earth   
6975   9394708  NaN                                           ghansoli   
7026   9799599  NaN                                              HKUST   
7055  10205276  NaN                                    Praha 10, Česko   
7079   5774750  NaN                                         Strickland   
7084   5793033  NaN                                             States   
7090   5813228  NaN                                        Cyber Space   
7213   6335837  NaN                                     中国Shanghai Shi   
7278   6630464  NaN                                       The Internet   
7286   6648456  NaN                                             Europe   
7312   6749539  NaN  In the desert eating dessert contemplating whe...   
7407   7296192  NaN  Universe, Virgo Supercluster, Local Group, Mil...   
7470   7663858  NaN                                  中国上海市Shanghai Shi   

                    creation_date  reputation  up_votes  down_votes  \
6163  2017-09-20 19:28:47.367 UTC         429       166          19   
6174   2017-10-22 03:32:06.66 UTC          93        10           0   
6180  2017-11-01 10:38:22.473 UTC          29        18           0   
6203  2017-11-27 14:22:04.717 UTC          55         9           0   
6211   2017-12-06 11:10:36.39 UTC          36         0           0   
6262  2018-03-23 00:48:57.613 UTC          55         2           0   
6309  2018-06-21 23:43:00.367 UTC          30         2           0   
6327   2018-08-11 10:33:01.33 UTC         632        74          10   
6356  2016-01-14 12:56:40.593 UTC          23         0           0   
6446   2016-04-08 09:17:07.11 UTC          96         5           0   
6508  2016-06-08 13:31:30.297 UTC          38         9           0   
6575  2016-08-11 19:49:52.637 UTC          65        17           0   
6603  2016-09-12 07:37:04.213 UTC         106         0           0   
6689  2016-12-30 16:45:07.183 UTC         166        27           4   
6788  2017-04-26 10:16:26.937 UTC          58         5           0   
6819  2017-06-19 09:59:39.717 UTC          36         9           0   
6860   2017-08-21 17:29:32.95 UTC         427         6          14   
6951   2018-01-08 12:48:32.05 UTC          84        17           0   
6975  2018-02-22 06:18:41.527 UTC          14         0           0   
7026  2018-05-16 10:43:15.847 UTC          43         2           0   
7055  2018-08-09 19:57:13.913 UTC         365        45           1   
7079  2016-01-11 16:52:58.473 UTC         857        27           0   
7084   2016-01-15 04:46:37.41 UTC        1242       118         656   
7090   2016-0

Looks that most of the unmatched entries are just people trying to be funny. "Earth" and "Moon" are quite popular locations, for instance.

In [44]:
world_cities[world_cities.city.str.lower().str.contains("\\bada\\b")]

city city_ascii      lat      lng country iso2 iso3 admin_name capital  \
2010  Ada        Ada  45.8027  20.1285  Serbia   RS  SRB        Ada   admin   

      population          id  
2010         NaN  1688687750

There's also a city called Ada in Serbia. :)